modules/ROOT/pages/test-mule-applications.adoc

```diff
@@ -0,0 +1,61 @@
+
+See the xref:munit::index.adoc[MUnit] documentation for more details.
+
+You can also test your Mule application's performance by profiling Mule runtime engine (Mule). This process can help you identify memory or performance-related issues in Mule applications or custom Mule extensions. To profile Mule, you need to load a Java Profiler into your on-premises Mule instance.
+
+See xref:profiling-mule.adoc[Profiling Mule] for more instructions.
+
+== Unit Tests
+
+The concept of unit tests may vary according to the programming paradigm that you are working on, but it always points to the same core concept: _to validate the correctness of an individual unit of source code_.
+
+In the context of unit testing, a unit of code is the smallest testable part of an application. What exactly constitutes the smallest testable part of an application depends on the application. When testing Mule applications, MuleSoft considers that the smallest testable part is a Mule flow or a sub-flow.
+
+Test cases must be independent of each other and it is the developer’s responsibility to ensure test inter-independency.
+The code you test might interact with other units of code or components, always consider this possibility when designing unit tests. A good unit test ensures the isolation of the unit of code being tested, to avoid mistaking failure in other components for failure in the unit of code being tested.
+
+To isolate the target unit of code, use tools such as the `mock-when` processor provided by MUnit.
+
+See xref:munit::mock-before-after-foreach-cookbook.adoc[Mocking a Message Before and Inside a Foreach Processor] for a usage example of the `mock-when` processor.
+
+== Integration Tests
+
+Units of code collaborate between them to create an actual application. You test individual units of code with unit tests, and you test how units of code collaborate between them with integration tests.
+
+The goal of an Integration test is _to validate that different units of code and modules work together as intended_.
+Integration tests must aggregate units of code modules that have already been successfully unit-tested. For this reason, always run your integration tests after your unit tests.
+
+Depending on the nature of your application, integration tests may require sandboxes, that is, environments for your application to connect to. When working with sandboxes, ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.
```

Comment:
```suggestion
Depending on the nature of your application, your integration tests might require a  _sandbox_ testing environment, which enables isolated testing evaluation outside of the working application environment. You must ensure that the state of the data in the sandbox is correct for the test to produce the intended result. Perform this verification before and after running the integration test.
```

Reply:
Please carefully evaluate the suggested rewording to ensure it does not change your intended meaning.
I wonder if we really need to define "sandbox"? Would this suffice:
```suggestion
Depending on the nature of your application, integration tests might require using a sandbox. If so, ensure that the state of the data in the sandbox is correct, both before and after running the integration test.
```
And doesn't this apply to any environment or test: that the state of the data be correct?